### PR TITLE
GEOPY-1811: b

### PR DESCRIPTION
**Pull request by domfournier**
**GEOPY-1811 - Use latest "Tx ID tx property" on ground EM metadata. **

### CODE REVIEW

File: geoh5py/objects/surveys/electromagnetics/base.py [hunk 1]
```diff
@@ -804,58 +810,22 @@ def default_input_types(self) -> list[str]:
...
-
-    @tx_id_property.setter
-    def tx_id_property(self, value: uuid.UUID | ReferencedData | np.ndarray | None):
-        if isinstance(value, uuid.UUID):
-            value = self.get_data(value)[0]
-
-        if isinstance(value, np.ndarray):
-            attributes = {
-                "values": value.astype(np.int32),
+    def _format_transmitter_ids(self, values, attributes):
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/base.py:813]: this function could be in BaseEMSurvey as you are using it for tipper too?
Just changing the first word of the value_map

File: geoh5py/objects/surveys/electromagnetics/tipper.py [hunk 2]
```diff
@@ -170,18 +167,27 @@ def default_units(self) -> list[str]:
         """
         return self.__UNITS
 
-    @property
-    def tx_id_property(self) -> ReferencedData | IntegerData | None:
-        """
-        Data link between the receiver and transmitter object.
+    def _format_transmitter_ids(self, values, attributes):
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/tipper.py:170]: See other comment about the same function

File: geoh5py/objects/surveys/electromagnetics/base.py [hunk 3]
```diff
@@ -291,9 +292,15 @@ def copy_complement(
             new_entity.tx_id_property is not None
             and self.complement.tx_id_property is not None
             and self.complement.tx_id_property.values is not None
+            and mask is not None
         ):
-            unique_ids = np.unique(self.complement.tx_id_property.values)
-            ids_mask = np.zeros(unique_ids.max() + 1, dtype=bool)
+            max_id = np.max(
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/base.py:297]: could almost be a function by itself for clarity?

as you want...

File: geoh5py/objects/surveys/electromagnetics/base.py [hunk 4]
```diff
@@ -291,9 +292,15 @@ def copy_complement(
             new_entity.tx_id_property is not None
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/base.py:292]: copy_complement has no docstring in BaseEm

**Review by MatthieuCMira — COMMENTED** [review]
small changes

File: geoh5py/objects/surveys/electromagnetics/base.py [hunk 5]
```diff
@@ -162,10 +162,11 @@ def add_validate_component_data(self, name: str, data_block: list | dict):
             )
 
         if isinstance(data_block, list):
-            assert np.all([entry.parent == self for entry in data_block]), (
-                f"The list of values provided for the component '{name}' "
-                f"must contain {FloatData} belonging to the target survey."
-            )
+            if not np.all([entry.parent == self for entry in data_block]):
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/base.py:165]: you could create a unitest to augment coverage for the error ;-)

File: geoh5py/objects/surveys/electromagnetics/base.py [hunk 6]
```diff
@@ -713,6 +715,10 @@ def tx_id_property(self, value: uuid.UUID | ReferencedData | np.ndarray | None):
         else:
             self.edit_em_metadata({"Tx ID tx property": getattr(value, "uid", None)})
 
+    def _format_transmitter_ids(self, _, attributes):
+        if self.complement is not None and self.complement.tx_id_property is not None:
+            attributes["entity_type"] = self.complement.tx_id_property.entity_type
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/base.py:720]: Not tested

File: geoh5py/objects/surveys/electromagnetics/base.py [hunk 7]
```diff
@@ -804,58 +810,22 @@ def default_input_types(self) -> list[str]:
         """Choice of survey creation types."""
         return self.__INPUT_TYPE
```
Comment (MatthieuCMira) [inline geoh5py/objects/surveys/electromagnetics/base.py:811]: increase coverage for this property

**Review by MatthieuCMira — COMMENTED** [review]
Some lines are not tested. 

As we are here, could you augment the coverage for errors?

File: geoh5py/objects/surveys/electromagnetics/base.py (hunk 1 above)
Comment (domfournier) [reply geoh5py/objects/surveys/electromagnetics/base.py:813]: Main difference is that LargeLoop and Tipper are referenced data with specific labels. Hence my thought of having slightly different methods on each.

File: geoh5py/objects/surveys/electromagnetics/tipper.py (hunk 2 above)
Comment (domfournier) [reply geoh5py/objects/surveys/electromagnetics/tipper.py:170]: We may need to clarify this in person

File: geoh5py/objects/surveys/electromagnetics/base.py (hunk 5 above)
Comment (domfournier) [reply geoh5py/objects/surveys/electromagnetics/base.py:165]: done

File: geoh5py/objects/surveys/electromagnetics/base.py (hunk 6 above)
Comment (domfournier) [reply geoh5py/objects/surveys/electromagnetics/base.py:720]: Done

File: geoh5py/objects/surveys/electromagnetics/base.py (hunk 7 above)
Comment (domfournier) [reply geoh5py/objects/surveys/electromagnetics/base.py:820]: Done

**Review by benk-mira — APPROVED** [review]

**Pull request merged**